File: frontend/src/lib/api/datasets.ts
import { apiGet, apiSend } from "@/lib/api/client";

export type VersionSummary = { id: string; version_number: number; file_size: number; row_count: number; column_count: number; duplicate_row_count: number; created_at: string };
export type DatasetSummary = { id: string; name: string; created_at: string; updated_at: string };
export type Source = { id: string; name: string; source_type: string; original_filename: string; status: string; created_at: string; updated_at: string; dataset: DatasetSummary; latest_version: VersionSummary; quality_issue_count: number };
export type QualityIssue = { id: string; column_name: string | null; code: string; severity: "INFO" | "WARNING" | "ERROR"; message: string; details: Record<string, unknown> };
export type ColumnProfile = { null_count: number; null_percentage: number; unique_count: number; numeric_min: number | null; numeric_max: number | null; numeric_mean: number | null; date_min: string | null; date_max: string | null; sample_values: unknown[] };
export type ProfileColumn = { id: string; name: string; position: number; inferred_type: string; generic_role: string; profile: ColumnProfile };
export type DatasetDetail = { id: string; source_id: string; name: string; source_type: string; original_filename: string; created_at: string; updated_at: string; latest_version: VersionSummary; quality_issue_count: number };
export type VersionDetail = VersionSummary & { columns: ProfileColumn[]; quality_issues: QualityIssue[] };
export type Preview = { dataset_id: string; version_id: string; columns: string[]; rows: Record<string, unknown>[]; limit: number };

export const listSources = () => apiGet<Source[]>("/api/v1/sources");
export const getDataset = (id: string) => apiGet<DatasetDetail>(`/api/v1/datasets/${id}`);
export const getVersion = (datasetId: string, versionId: string) => apiGet<VersionDetail>(`/api/v1/datasets/${datasetId}/versions/${versionId}`);
export const getPreview = (id: string, limit = 50) => apiGet<Preview>(`/api/v1/datasets/${id}/preview?limit=${limit}`);
export function uploadSource(file: File) { const body = new FormData(); body.append("file", file); return apiSend<Source>("/api/v1/sources/upload", { method: "POST", body }); }

File: backend/app/schemas/datasets.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict


class QualityIssueRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    column_name: str | None
    code: str
    severity: str
    message: str
    details: dict[str, Any]


class FieldProfileRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    null_count: int
    null_percentage: float
    unique_count: int
    numeric_min: float | None
    numeric_max: float | None
    numeric_mean: float | None
    date_min: str | None
    date_max: str | None
    sample_values: list[Any]


class ColumnRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    name: str
    position: int
    inferred_type: str
    generic_role: str
    profile: FieldProfileRead


class VersionSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    version_number: int
    file_size: int
    row_count: int
    column_count: int
    duplicate_row_count: int
    created_at: datetime


class DatasetSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    name: str
    created_at: datetime
    updated_at: datetime


class SourceRead(BaseModel):
    id: str
    name: str
    source_type: str
    original_filename: str
    status: str
    created_at: datetime
    updated_at: datetime
    dataset: DatasetSummary
    latest_version: VersionSummary
    quality_issue_count: int


class DatasetDetail(BaseModel):
    id: str
    source_id: str
    name: str
    source_type: str
    original_filename: str
    created_at: datetime
    updated_at: datetime
    latest_version: VersionSummary
    quality_issue_count: int


class VersionDetail(VersionSummary):
    columns: list[ColumnRead]
    quality_issues: list[QualityIssueRead]


class ProfileRead(BaseModel):
    dataset_id: str
    version: VersionDetail


class PreviewRead(BaseModel):
    dataset_id: str
    version_id: str
    columns: list[str]
    rows: list[dict[str, Any]]
    limit: int
